pallets/erc20-xcm-bridge/src/lib.rs

```diff
@@ -162,7 +161,11 @@ pub mod pallet {
 			let to = T::AccountIdConverter::convert_ref(to)
 				.map_err(|()| MatchError::AccountIdConversionFailed)?;
 
-			Self::erc20_transfer(contract_address, from, to, amount)?;
+			// We perform the evm transfers in a storage transaction to ensure that if it fail
+			// any contract storage changes are rolled back.
+			frame_support::storage::with_storage_layer(|| {
```

Comment:
We already execute the evm in a db transaction, why starting a new one?

Regarding that one thing I just noticed is that `with_storage_layer` uses a nested transactional limit, which would `Err` if that limit is reached. We should take the limit into account when manually calling `start_transaction` etc in `SubstrateStackSubstate` in frontier.

Reply:
> We already execute the evm in a db transaction, why starting a new one?

Not necessarily, if we are in the context of a precompile, the subcall is not done in a db transaction.

We have to be sure at the pallet level that any change will be rolled back in case of a problem, we can't rely on the external implementation of the runner.

Reply:
Any evm subcall is done in a new db transaction when calling [enter_substate](https://github.com/rust-blockchain/evm/blob/master/src/executor/stack/executor.rs#L929).

Reply:
Yes, but the rollback will only be done if the execution throws an error, but here we want to rollback in some cases even if the evm execution doesn't return any error (for example when the contract returns false).